test/vegeta/pacers.go

```diff
@@ -0,0 +1,225 @@
+}
+
+// invalid tests the constraints documented in the steadyUpPacer struct definition.
+func (sup *steadyUpPacer) invalid() bool {
+	return sup.UpDuration <= 0 || sup.minHitsPerNs <= 0 || sup.maxHitsPerNs <= sup.minHitsPerNs
+}
+
+// Pace determines the length of time to sleep until the next hit is sent.
+func (sup *steadyUpPacer) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	if sup.invalid() {
+		// If pacer configuration is invalid, stop the attack.
+		return 0, true
+	}
+
+	expectedHits := sup.hits(elapsedTime)
+	if elapsedHits < uint64(expectedHits) {
+		// Running behind, send next hit immediately.
+		return 0, false
+	}
+
+	// Re-arranging our hits equation to provide a duration given the number of
+	// requests sent is non-trivial, so we must solve for the duration numerically.
+	// math.Round() added here because we have to coerce to int64 nanoseconds
+	// at some point and it corrects a bunch of off-by-one problems.
+	nsPerHit := 1 / sup.hitsPerNs(elapsedTime)
+	hitsToWait := float64(elapsedHits+1) - expectedHits
+	nextHitIn := time.Duration(nsPerHit * hitsToWait)
+
+	// If we can't converge to an error of <1e-3 within 10 iterations, bail.
+	// This rarely even loops for any large Period if hitsToWait is small.
+	for i := 0; i < 10; i++ {
+		hitsAtGuess := sup.hits(elapsedTime + nextHitIn)
+		err := float64(elapsedHits+1) - hitsAtGuess
+		if math.Abs(err) < 1e-3 {
+			return nextHitIn, false
+		}
+		nextHitIn = time.Duration(float64(nextHitIn) / (hitsAtGuess - float64(elapsedHits)))
+	}
+
+	return nextHitIn, false
+}
+
+// hits returns the number of expected hits for this pacer during the given time.
+func (sup *steadyUpPacer) hits(t time.Duration) float64 {
+	if t <= 0 || sup.invalid() {
+		return 0
+	}
+
+	// If t is smaller than the UpDuration, calculate the hits as a trapezoid.
+	if t <= sup.UpDuration {
+		curtHitsPerNs := sup.hitsPerNs(t)
+		return (curtHitsPerNs + sup.minHitsPerNs) / 2.0 * float64(t)
+	}
+
+	// If t is larger than the UpDuration, calculate the hits as a trapezoid + a rectangle.
+	upHits := (sup.maxHitsPerNs + sup.minHitsPerNs) / 2.0 * float64(sup.UpDuration)
+	steadyHits := sup.maxHitsPerNs * float64(t-sup.UpDuration)
+	return upHits + steadyHits
+}
+
+// hitsPerNs returns the attack rate for this pacer at a given time.
+func (sup *steadyUpPacer) hitsPerNs(t time.Duration) float64 {
+	if t <= sup.UpDuration {
+		return sup.minHitsPerNs + float64(t)*sup.slope
+	}
+
+	return sup.maxHitsPerNs
+}
+
+// hitsPerNs returns the attack rate this ConstantPacer represents, in
+// fractional hits per nanosecond.
+func hitsPerNs(cp vegeta.ConstantPacer) float64 {
+	return float64(cp.Freq) / float64(cp.Per)
+}
+
+// combinedPacer is a pacer that combines multiple pacers and runs them sequentially when being used for attack.
+type combinedPacer struct {
+	// Pacers is a list of pacers that will be used sequentially for attack, must be non-empty.
+	Pacers []vegeta.Pacer
```

Comment:
Any reason for those to be public?

Reply:
Actually no.. Changed them to be private.